sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/batch/implementation/ItemBatchOperation.java

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.batch.implementation;
+
+import com.azure.cosmos.implementation.*;
+import com.azure.cosmos.implementation.apachecommons.lang.StringUtils;
+import com.azure.cosmos.implementation.directconnectivity.WFConstants;
+import com.azure.cosmos.models.PartitionKey;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+import static com.azure.cosmos.batch.implementation.BatchRequestResponseConstant.*;
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkArgument;
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+/**
+ * Represents an operation on an item which will be executed as part of a batch request on a container.
+ */
+public final class ItemBatchOperation<TResource> implements AutoCloseable {
+
+    private TResource resource;
+    private String materialisedResource;
+
+    private ItemBatchOperationContext context;
+    private String id;
+    private int operationIndex;
+    private PartitionKey partitionKey;
+    private String partitionKeyJson;
+    private final OperationType operationType;
+    private RequestOptions requestOptions;
+
+    private ItemBatchOperation(
+        final OperationType operationType,
+        final int operationIndex,
+        final PartitionKey partitionKey,
+        final String id,
+        final TResource resource,
+        final RequestOptions requestOptions) {
+
+        checkArgument(operationIndex >= 0, "expected operationIndex >= 0, not %s", operationIndex);
+        checkNotNull(operationType, "expected non-null operationType");
+
+        this.operationType = operationType;
+        this.operationIndex = operationIndex;
+        this.partitionKey = partitionKey;
+        this.id = id;
+        this.resource = resource;
+        this.requestOptions = requestOptions;
+    }
+
+    // TODO(rakkuma): Similarly for hybrid row, operation needs to be written in Hybrid row.
+    public static JsonSerializable writeOperation(ItemBatchOperation<?> operation) {
+        JsonSerializable jsonSerializable = new JsonSerializable();
+
+        jsonSerializable.set(FIELD_OPERATION_TYPE, BatchExecUtils.getStringOperationType(operation.getOperationType()));
+
+        if (operation.getPartitionKey() != null && StringUtils.isNotEmpty(operation.getPartitionKey().toString())) {
```

Comment:
```operation.getPartitionKey().toString()```
will serialize the pk to json and is costly, is that intentional?

Reply:
No. Will update it to get it only once. Thanks.